### PR TITLE
ESLintBear.py: Add settings from eslint

### PR DESCRIPTION
**Pull request by anctartica**
Add document rules from eslint

Closes https://github.com/coala/coala-bears/issues/1951

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by anctartica** [comment]
I am working on it!

**Comment by anctartica** [comment]
all the tests are passing when i run pytest locally on my computer even when i add invalid files to valid_files in verify_local_bear @newbazz
I believe my tests are being skipped it shows ssss next to them. What is the reason for this?

**Comment by anctartica** [comment]
@newbazz there are no rules/settings in standard that can be changed...
I referred to this: 
https://standardjs.com/#javascript-style-guide-with-linter--automatic-code-fixer
https://standardjs.com/#i-disagree-with-rule-x-can-you-change-it
It comes with predefined rules. 
So i believe there are no settings that the user can configure
https://github.com/coala/coala-bears/issues/1951#issuecomment-322638993
Or am i interpreting this completely wrong?

**Comment by newbazz** [comment]
@anctartica but you have coverage problem, you need to add tests in `JSSBeartests.py` ;).

**Comment by anctartica** [comment]
@newbazz i understand that i don't have tests so far... but I am saying that it is not possible to add all those settings that I am trying to add to the JSStandardBear so, we won't need tests for coverage

**Comment by newbazz** [comment]
I went through the link you send, and still you can try some settings to implement ;P
cc @Makman2

**Comment by anctartica** [comment]
ohh okay... I am trying to add these few settings but there is no way to specify a configuration file on the command line in standard.. we would have to add these individually like `--parser babel-eslint --plugin flowtype`. I am not sure how i would achieve this @newbazz

**Comment by anctartica** [comment]
I think I have figured it out

File: bears/js/JSStandardBear.py [hunk 1]
```diff
@@ -24,5 +24,16 @@ class JSStandardBear:
...
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         parser: str = '',
+                         globals: str= '',
+                         plugin: str = ''):
+        args = ('--verbose', filename)
+        if parser != '':
```
Comment (newbazz) [inline bears/js/JSStandardBear.py]: you can right it as `if parser:` right.

File: bears/js/JSStandardBear.py [hunk 2]
```diff
@@ -24,5 +24,16 @@ class JSStandardBear:
...
+            args += ('--parser', parser)
+        if globals != '':
```
Comment (newbazz) [inline bears/js/JSStandardBear.py]: why `globals` is in blue?

Comment (anctartica) [reply bears/js/JSStandardBear.py]: I am not exactly sure. But it does refer to the value that gets passed to the globals parameter

File: bears/js/JSStandardBear.py (hunk 1 above)
Comment (anctartica) [reply bears/js/JSStandardBear.py]: done!

File: bears/js/JSStandardBear.py (hunk 2 above)
Comment (anctartica) [reply bears/js/JSStandardBear.py]: globals is built-in name and we are masking it... that is why

**Comment by anctartica** [comment]
are there any other changes to be made? @newbazz

**Comment by newbazz** [comment]
LGTM ;)

File: bears/js/JSStandardBear.py [hunk 3]
```diff
@@ -24,5 +25,17 @@ class JSStandardBear:
     CAN_FIX = {'Formatting'}
     SEE_MORE = 'https://standardjs.com/rules.html'
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         parser: str = '',
+                         globals: typed_list(str)= (),
+                         plugin: str = ''):
```
Comment (Makman2) [inline bears/js/JSStandardBear.py]: no spaces around `=`

Comment (Makman2) [inline bears/js/JSStandardBear.py]: The idea is actually to implement the rules of `standard` directly, the user shouldn't deal with plugins and stuff, but just say `lint-xyz = True`.

File: bears/js/JSStandardBear.py [hunk 4]
```diff
@@ -24,5 +25,17 @@ class JSStandardBear:
     CAN_FIX = {'Formatting'}
     SEE_MORE = 'https://standardjs.com/rules.html'
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         parser: str = '',
+                         globals: typed_list(str)= (),
```
Comment (Makman2) [inline bears/js/JSStandardBear.py]: shadows built-in name `globals`

**Comment by Makman2** [comment]
unack 7643560

**Comment by Makman2** [comment]
Actually it seems the issue description is a bit bad... Anyway please implement settings in the manner like other bears ;)

File: bears/js/JSStandardBear.py (hunk 3 above)
Comment (anctartica) [reply bears/js/JSStandardBear.py]: so as I understand here `lint-xyz` would be the name of a parser? @Makman2

Comment (Makman2) [reply bears/js/JSStandardBear.py]: no the capability of a parser, like `remove_unnecessary_spaces` or `disallow_strings` so (if that would exist).

Comment (Makman2) [reply bears/js/JSStandardBear.py]: The idea of coala is to abstract those things like "parsers", "plugins" etc etc away. The user has our bears, which accept settings that directly correlate to a certain analyis. The user shall not deal with internals of tools like `standard`, i.e. that it uses plugins and parsers and globals.

Comment (anctartica) [reply bears/js/JSStandardBear.py]: so i need to add settings for the user according to the capabilities of the different parsers that can be used?  @Makman2

Comment (Makman2) [reply bears/js/JSStandardBear.py]: exactly ;)

**Comment by anctartica** [comment]
I don't know understand why TextLintBear tests are failing because of my code for JSStandardBear. @Makman2

**Comment by Makman2** [comment]
You've hit https://github.com/coala/coala-bears/issues/2191 . A PR is on the way, but it probably will take a bit still because we have additional checks to prevent the patches that are necessary.

File: bears/js/JSStandardBear.py [hunk 5]
```diff
@@ -16,13 +17,36 @@ class JSStandardBear:
     """
 
     LANGUAGES = {'JavaScript', 'JSX'}
-    REQUIREMENTS = {NpmRequirement('standard', '10')}
+    REQUIREMENTS = {NpmRequirement('standard', '10.0.3'),
+                    NpmRequirement('babel-eslint', '8'),
+                    NpmRequirement('eslint-plugin-flowtype', '2.41.0'),
+                    NpmRequirement('typescript-eslint-parser', '12.0.0'),
+                    NpmRequirement('eslint-plugin-typescript', '0.8.1')}
```
Comment (Makman2) [inline bears/js/JSStandardBear.py]: those are `eslint` requirements, they don't affect `standard`, do they?

File: bears/js/JSStandardBear.py [hunk 6]
```diff
@@ -16,13 +17,36 @@ class JSStandardBear:
...
     SEE_MORE = 'https://standardjs.com/rules.html'
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    def create_arguments(self, filename, file, config_file,
+                         parser: str='',
+                         globalVars: typed_list(str)=(),
+                         plugin: str='',
+                         typescript: bool=False,
+                         flow: bool=False):
```
Comment (Makman2) [inline bears/js/JSStandardBear.py]: Like said in the comment before, we don't want settings like, `plugin` or `parser`. See other bears how settings shall look like.

File: bears/js/JSStandardBear.py (hunk 5 above)
Comment (anctartica) [reply bears/js/JSStandardBear.py]: we do require them for checking typescript and flow with standard

File: bears/js/JSStandardBear.py (hunk 6 above)
Comment (anctartica) [reply bears/js/JSStandardBear.py]: the `parser` option will be removed

File: bears/js/JSStandardBear.py (hunk 5 above)
Comment (anctartica) [reply bears/js/JSStandardBear.py]: yes... https://standardjs.com/index.html#flow

File: bears/js/JSStandardBear.py [hunk 7]
```diff
@@ -16,13 +17,34 @@ class JSStandardBear:
...
     CAN_DETECT = {'Syntax'}
     CAN_FIX = {'Formatting'}
     SEE_MORE = 'https://standardjs.com/rules.html'
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    def create_arguments(self, filename, file, config_file,
+                         globalVars: typed_list(str)=(),
+                         typescript: bool=False,
+                         flow: bool=False):
```
Comment (Makman2) [inline bears/js/JSStandardBear.py]: Documentation missing ;)

File: bears/js/JSStandardBear.py [hunk 8]
```diff
@@ -16,13 +17,34 @@ class JSStandardBear:
...
+        args = ('--verbose', filename,)
+        if flow and typescript:
+            self.err('The javascript variants flow and typescript are ',
+                     'both given but cannot be used together.')
```
Comment (Makman2) [inline bears/js/JSStandardBear.py]: Could you check if coala doesn't cache results if those two settings conflict, and afterwards they are used correctly?

File: tests/js/JSStandardBearTest.py [hunk 9]
```diff
@@ -77,3 +95,31 @@
                                                       bad_file_undef,
                                                       bad_file_ifelse,
                                                       bad_file_func_name,))
+
+
+jsstandardBearTsTest = verify_local_bear(
```
Comment (Makman2) [inline tests/js/JSStandardBearTest.py]: Please adhere to the correct casing + Don't shorten parts of the name ;) Ts --> Typescript

File: tests/js/JSStandardBearTest.py [hunk 10]
```diff
@@ -77,3 +95,31 @@
...
+    valid_files=(test_file_flow,),
+    settings={'flow': True})
+
+jsstandardBearGlobalVarTest = verify_local_bear(
+    JSStandardBear,
+    invalid_files=(),
+    valid_files=(test_file_globalVar,),
+    settings={'globalVars': 'x,y,z'})
+
+jsstandardBearTsFlowTest = verify_local_bear(
```
Comment (Makman2) [inline tests/js/JSStandardBearTest.py]: same here

File: bears/js/JSStandardBear.py [hunk 11]
```diff
@@ -16,13 +17,34 @@ class JSStandardBear:
...
     CAN_FIX = {'Formatting'}
     SEE_MORE = 'https://standardjs.com/rules.html'
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    def create_arguments(self, filename, file, config_file,
+                         globalVars: typed_list(str)=(),
+                         typescript: bool=False,
+                         flow: bool=False):
+        args = ('--verbose', filename,)
```
Comment (Makman2) [inline bears/js/JSStandardBear.py]: if you append so much, rather use a list :+1:

**Comment by Makman2** [comment]
Optional: Maybe describe quickly in the commit body what settings / what kind of settings you've added?

**Comment by Makman2** [comment]
Also are these all available settings?

**Comment by Makman2** [comment]
unack 1068a01326377a98598a718bf3c779fa1911f69f

File: bears/js/JSStandardBear.py (hunk 8 above)
Comment (anctartica) [reply bears/js/JSStandardBear.py]: I am not sure how i would do this... do you want me to create a test or check locally on my machine?

**Comment by anctartica** [comment]
I have added all the available settings now @Makman2

File: bears/js/JSStandardBear.py [hunk 12]
```diff
@@ -14,15 +15,65 @@ class JSStandardBear:
     No .eslintrc, .jshintrc, or .jscsrc files to manage.
     It just works.
     """
-
     LANGUAGES = {'JavaScript', 'JSX'}
-    REQUIREMENTS = {NpmRequirement('standard', '10')}
+    REQUIREMENTS = {NpmRequirement('standard', '10'),
+                    NpmRequirement('babel-eslint', '8'),
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: can this be changed to `8.0` ? that will fix the problem with Travis.

File: bears/js/JSStandardBear.py [hunk 13]
```diff
@@ -14,15 +15,65 @@ class JSStandardBear:
...
     SEE_MORE = 'https://standardjs.com/rules.html'
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    def create_arguments(self, filename, file, config_file,
+                         globalVars: typed_list(str)=(),
+                         typescript: bool=False,
+                         flow: bool=False,
+                         env: str='',
+                         markdown: bool=False,
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: nope; this should take a `language: Language`, accepting HTML/Markdown, and add the necessary parameters behind the scenes.

TypeScript/flow should also be defined languages, IMO, and should be handled the same way, but they are at least variants of JavaScript so maybe others have an alternative opinion about allowing those to be parameters. 

You may need to define some of those languages in coala core's coalang.

You might like to split this patch into two, one adding `env` and `globalVars`, since those are less complicated.

Comment (anctartica) [reply bears/js/JSStandardBear.py]: okay yes that would be much cleaner rather than having all these options. Currently flow and markdown are not included in the coalang. Should I simply make a PR to add these two languages first and then finish the work here? @jayvdb

**Comment by jayvdb** [comment]
You can probably get this unblocked quicker by only doing markdown.  I expect flow is going to take a bit longer.

**Comment by anctartica** [comment]
The logs show that ``Markdown` is not a valid language name or not recognized by coala.` But all the tests are passing locally because the changes (adding Markdown definition) had taken effect in the `virtualenv` where I was working. Seems like this is not the case with the CI. How can this be resolved?

File: bears/js/JSStandardBear.py [hunk 14]
```diff
@@ -14,15 +16,54 @@ class JSStandardBear:
...
     """
+    supported_languages = ['Markdown', 'html', 'TypeScript']
 
     LANGUAGES = {'JavaScript', 'JSX'}
-    REQUIREMENTS = {NpmRequirement('standard', '10')}
+    REQUIREMENTS = {NpmRequirement('standard', '10'),
+                    NpmRequirement('typescript-eslint-parser', '12.0.0'),
+                    NpmRequirement('eslint-plugin-typescript', '0.8.1'),
+                    NpmRequirement('eslint-plugin-markdown', '1.0.0-beta.6'),
+                    NpmRequirement('eslint-plugin-html', '3.2.2'), }
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: a trailing `,` is useless if the trailing `}` is on the same line.

File: bears/js/JSStandardBear.py [hunk 15]
```diff
@@ -14,15 +16,54 @@ class JSStandardBear:
...
     CAN_DETECT = {'Syntax'}
     CAN_FIX = {'Formatting'}
     SEE_MORE = 'https://standardjs.com/rules.html'
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    def create_arguments(self, filename, file, config_file,
+                         language: language = language('JavaScript'),
+                         globalVars: typed_list(str) = (),
+                         env: str = ''):
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: use trailing `,` and put `):` on a different line

File: bears/js/JSStandardBear.py [hunk 16]
```diff
@@ -14,15 +16,54 @@ class JSStandardBear:
...
     LICENSE = 'AGPL-3.0'
     CAN_DETECT = {'Syntax'}
     CAN_FIX = {'Formatting'}
     SEE_MORE = 'https://standardjs.com/rules.html'
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    def create_arguments(self, filename, file, config_file,
+                         language: language = language('JavaScript'),
+                         globalVars: typed_list(str) = (),
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: `globalVars` is not PEP8 compatible.

File: bears/js/JSStandardBear.py [hunk 17]
```diff
@@ -14,15 +16,54 @@ class JSStandardBear:
     No .eslintrc, .jshintrc, or .jscsrc files to manage.
     It just works.
     """
+    supported_languages = ['Markdown', 'html', 'TypeScript']
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: why is this here?

Shouldnt these be in `LANGUAGES` ?

File: bears/js/JSStandardBear.py [hunk 18]
```diff
@@ -16,13 +18,52 @@ class JSStandardBear:
...
+        :param global_vars:
+            The list of global variables in the code
+        :param language:
+            Set to ``Javascript`` by default. To check javascript in either
+            markdown or html specify ``Markdown`` or ``html``. Use
+            ``Typescript`` to lint typescript.
+        :param env:
+            To support global variables of certain environments. The complete
+            list of supported environments can be found here
+            https://eslint.org/docs/user-guide/configuring.HTML#specifying-environments
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: 404 with the `.HTML`

doc now at

https://eslint.org/docs/user-guide/configuring#specifying-environments

Note this settings should accept a typed_list(str) as it has many values.

might be useful to give this setting a more meaningful name, like `javascript_global_variable_environments` .  
We have a few ESLint based linters, and presumably this setting should be common to them all.
JSHint and JSLint also have a very similar concept.

Alternatively , we can name it the same as the inline comment syntax :  eslint-env or eslint_env

File: bears/js/JSStandardBear.py [hunk 19]
```diff
@@ -16,13 +18,52 @@ class JSStandardBear:
...
-        return (filename, '--verbose')
+    def create_arguments(self, filename, file, config_file,
+                         language: language = language('JavaScript'),
+                         global_vars: typed_list(str) = (),
+                         env: str = '',
+                         ):
+        """
+        :param global_vars:
+            The list of global variables in the code
+        :param language:
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: language should appear before global_vars here in the docstring.

**Comment by jayvdb** [comment]
Tests failing.

**Comment by anctartica** [comment]
it shows `Language `Markdown` is not a valid language name or not recognized by coala.`
It requires an updated version of coala

File: bears/js/JSStandardBear.py [hunk 20]
```diff
@@ -16,13 +18,53 @@ class JSStandardBear:
...
 
-    def create_arguments(self, filename, file, config_file):
-        return (filename, '--verbose')
+    def create_arguments(self, filename, file, config_file,
+                         language: language = language('JavaScript'),
+                         global_vars: typed_list(str) = (),
+                         eslint_env: typed_list(str) = (),
+                         ):
+        """
+        :param language:
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: all of these plugins, and all of these options, all apply to `eslint`.  It is going to be wierd if JSStandard bear is wildly different from ESLint, e.g. doesnt use `babel-eslint`.

Soooo ... move allllll of this stuff to `ESLintBear`.

Very likely, all of your tests also are almost applicable for `ESLintBear`.

Dont want to write extra tests?  No worries.

Make StandardBear inherit from ESLintBear.
All of your current tests for StandardBear will cover the lines of ESLintBear that are moved there.

`standard` ignores the `--stdin` and `--stdin-filename` args which ESLintBear users.  Create an upstream bug about those -- we want those.  json is yummier than plain text.

But we can workaround it atm:

`JSStandardBear.create_arguments` would call `super` `ESLintBear.create_arguments` , and then add the filename to the end of the args.

`JSStandardBear.process_output` would skip `ESLint.process_output` and call `Bear.process_output`

It also ignores `--no-color` `--no-ignore` but meh .. we dont need those anyway.

`ASCIINEMA_URL` needs to be explicitly set to `None` to override `ESLint`.

File: bears/js/JSStandardBear.py [hunk 21]
```diff
@@ -16,13 +18,53 @@ class JSStandardBear:
...
+            args += ('--plugin', 'markdown',)
+        elif 'Typescript' in language:
+            args += ('--parser', 'typescript-eslint-parser',
+                     '--plugin', 'typescript')
+        elif 'HTML' in language:
+            args += ('--plugin', 'html')
+        elif 'Javascript' not in language:
+            self.err(
+                'Language needs to be either Markdown, HTML, Typescript '
+                'or Javascript. Assuming Javascript.')
```
Comment (wisn) [inline bears/js/JSStandardBear.py]: I want to give a review after your decision toward @jayvdb [review](https://github.com/coala/coala-bears/pull/2230/files#r204232076). However, sorry for bring this up, could you write a proper name? Like, change `Javascript` to `JavaScript` instead. This also applies to TypeScript. All of them please.

File: bears/js/JSStandardBear.py (hunk 20 above)
Comment (anctartica) [reply bears/js/JSStandardBear.py]: `--stdin` and `--stdin-filename` are already supported I believe. I tested by running `cat test.js | standard --stdin --stdin-filename=test.js`. The documentation links to [standard-json](https://www.npmjs.com/package/standard-json) package for formatting to json output.

Otherwise I generally agree with your idea seeing that standard is only using rules from eslint. I should be able to move my code from here directly to `EslintBear`.

File: bears/js/JSStandardBear.py [hunk 22]
```diff
@@ -16,13 +18,53 @@ class JSStandardBear:
...
+            markdown or html specify ``Markdown`` or ``html``. Use
+            ``Typescript`` to lint typescript.
+        :param global_vars:
+            The list of global variables in the code
+        :param eslint_env:
+            To support global variables of certain environments. The complete
+            list of supported environments can be found here
+            https://eslint.org/docs/user-guide/configuring#specifying-environments
+        """
+        args = ('--verbose', filename,)
```
[the remark this answers is not in the document]
Comment (anctartica) [reply bears/js/JSStandardBear.py]: created here https://github.com/coala/coala/issues/5659

File: bears/js/JSStandardBear.py [hunk 23]
```diff
@@ -16,13 +18,53 @@ class JSStandardBear:
     """
 
     LANGUAGES = {'JavaScript', 'JSX'}
```
Comment (jayvdb) [inline bears/js/JSStandardBear.py]: note this will also need to change for both bears, as 'JS in Markdown' is still just Markdown.

File: bears/js/JSStandardBear.py (hunk 20 above)
Comment (jayvdb) [reply bears/js/JSStandardBear.py]: Assigned to you.  Please just do it in this PR.

And as *everything* in this PR can be put in `ESLintBear`, there will be another commit in this PR for adapting JSStandardBear to inheriting from ESLintBear.  I want to see those two together, so we get it right.

Otherwise we'll change ESLintBear incorrectly, and wont be able to build JSStandardBear on top.

File: tests/js/ESLintBearTest.py [hunk 24]
```diff
@@ -110,3 +184,49 @@ def test_lint_config_file(self):
...
+        self.section.language = Language['Python']
+        self.check_validity(self.uut, test_file)
+        while not self.queue.empty():
+            message = self.queue.get()
+            msg = ('Language needs to be either Markdown, HTML, Typescript '
+                   'or Javascript. Assuming Javascript.')
+            if message.message == msg:
+                break
+        else:
+            assert False, 'Message not found'
```
Comment (jayvdb) [inline tests/js/ESLintBearTest.py]: this one is currently failing.

to help with debugging, and also good for if it ever breaks, collect all of the messages encountered, and include them in the assertion message here.

or, collect all of the messages encountered, and use

`self.assertIn(expected_message, messages_encountered)`

That may help us see what went wrong.

**Comment by jayvdb** [comment]
ack https://github.com/coala/coala-bears/pull/2230/commits/e62d16f8315e4dcd10c9c57af3ab3e5973b10d73

**Comment by jayvdb** [comment]
@gitmate-bot ff
